### ISSUE TITLE
where is ODPS datahub table storaged in?

### ISSUE THREAD

**Issue by lchh0917** [aliyun/aliyun-odps-ogg-plugin#2]
Dear all:
        When I use aliyun-odps-ogg-plugin to complete realtime data duplicate, just only see the flow record table, which auto created by the ogg-odps-table-1.0.0.jar according this command: 
cd %YOUR_aliyun-odps-ogg-plugin_DIRECTORY%/datahub_handler/table_tool/
java -jar ogg-odps-table-1.0.0.jar CREATE %YOUR_OGG_DIRECTORY%/dirprm/extdatahub.properties

the flow record table data like this: 

**odps@ prj_longshine_001>desc tcustmer;**

+------------------------------------------------------------------------------------+
| Owner: RAM$cis_crm_ali1:changhua.lu | Project: prj_longshine_001
| TableComment:                                                                      |
+------------------------------------------------------------------------------------+
| CreateTime:               2016-10-14 15:46:22                                      |
| LastDDLTime:              2016-10-14 15:46:22                                      |
| LastModifiedTime:         2016-10-17 15:17:37                                      |
+------------------------------------------------------------------------------------+
| InternalTable: YES      | Size: 5488                                               |
+------------------------------------------------------------------------------------+
| Native Columns:                                                                    |
+------------------------------------------------------------------------------------+
| Field           | Type       | Label | Comment                                     |
+------------------------------------------------------------------------------------+
| optype          | string     |       |                                             |
| readtime        | string     |       |                                             |
| state           | string     |       |                                             |
| state_before    | string     |       |                                             |
| state_after     | string     |       |                                             |
+------------------------------------------------------------------------------------+
| Partition Columns:                                                                 |
+------------------------------------------------------------------------------------+
| pt              | string     |                                                     |
+------------------------------------------------------------------------------------+

OK

**odps@ prj_longshine_001>select * from tcustmer;**

ID = 20161017071716763g1ykp2i8
Log view:
http://logview.odps.aliyun.com/logview/?h=http://service.odps.aliyun.com/api&p=prj_lon……
+--------+----------+-------+--------------+-------------+----+
| optype | readtime | state | state_before | state_after | pt |
+--------+----------+-------+--------------+-------------+----+
| DELETE(3) | 2016-10-17 15:16:24.936 | NULL  | NULL         | NULL        | 161017 |
| INSERT(5) | 2016-10-17 15:10:40.050 | NULL  | NULL         | CO          | 161017 |
| INSERT(5) | 2016-10-17 15:10:40.051 | NULL  | NULL         | WA          | 161017 |
| INSERT(5) | 2016-10-17 15:11:32.531 | NULL  | NULL         | WA          | 161017 |
| INSERT(5) | 2016-10-17 15:11:32.532 | NULL  | NULL         | CO          | 161017 |
| INSERT(5) | 2016-10-17 14:49:48.841 | NULL  | NULL         | CO          | 161017 |
+--------+----------+-------+--------------+-------------+----+

**so, where is the replicated table data storaged in ODPS?**

**Comment by tianliplus** [comment]
This ogg plugin only send flow records to ODPS. You need to write SQL job to produce replicated table from historical table and flow records.

**Comment by lchh0917** [comment]
Hi,
what's the "historical table" you mentioned, can you give a typical example to refer? after all, I can't find a full scene case to learn.
Thanks!

**Comment by lchh0917** [comment]
And one thing more, I test both Oracle and MySQL as source DB, found aliyun-odps-ogg-plugin just only supported Oracle as source DB, oppositely MySQL occured error, so that's truly?

**Comment by tianliplus** [comment]
@lchh0917 
Assuming we have two tables, oracle table `t_oracle` and ODPS table `t_odps`, with the same data. And we want to synchronize the incrementation of `t_oracle` to `t_odps`. Since ODPS table is "immutable", it's very hard to implement real-time synchronization between the two tables. Therefore, we choose another way to implement a daily synchronization of them. First, we use this plugin to store flow records of `t_oracle` in another ODPS table `t_logs`, usually one partition per day. Then, we can produce the "synchronized" table `t_odps_20161018` from `t_odps` and `t_logs/pt=20161018` using ODPS SQL. By assigning this ODPS SQL to daily task, we can produce `t_odps_20161019` from `t_odps_20161018` and `t_logs/pt=20161019`, produce `t_odps_20161020` ... , and so on.

**Comment by tianliplus** [comment]
Yes, only oracle DB has been well tested. There may be some bugs with other DBs. 
Would you please post the error messages? I'll see if there is anything I can do.
